db: api: service: introduce system.client_routes table and related API endpoints

andrzej-jackowski-scylladb opened the pull request:
`system.client_routes` is a system table that sets the target address and ports for each `host_id`, for one or more connection (e.g., Private Link) represented by `connection_id`. Cloud will write the table via REST, and drivers will read it via CQL to override values obtained from `system.local` and `system.peers`.

This patch series contains:
 - Introduction of `CLIENT_ROUTES` feature flag.
 - Implementation of raft-based `system.client_routes` table
 - Implementation of `v2/client-routes` POST/DELETE/GET endpoints
 - Implementation of new `CLIENT_ROUTES_CHANGE` event that is sent to drivers when `system.client_routes` is changed
 - New tests that verifies the aforementioned features

Ref: scylladb/scylla-enterprise#5699

For now, no automatic backport. However, the changes are planned to be release on `2025.4` either as a backport or a private build.

andrzej-jackowski-scylladb commented:
v2:
 - Added missing `@skip_mode('release', 'error injections are not supported in release mode')` decorator to a test with injected errors

andrzej-jackowski-scylladb commented:
v3:
 - Removed `rack` and `datacenter` columns and related JSON fields
 - Changed GET endpoint to use CL=ONE instead of QUORUM
 - Fixed handling null terminators in JSON strings
 - Added `test_client_routes_lost_quorum` and `test_client_routes_null_terminators`

andrzej-jackowski-scylladb commented:
> #### Failed Tests (10/78756):
> * [test_events.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/debug___test_py___debug___test_events_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.debug.1)
> * [test_events.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/debug___test_py___debug___test_events_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.debug.2)
> * [test_events.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/debug___test_py___debug___test_events_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.debug.3)
> * [test_events.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/dev___test_py___dev___test_events_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.dev.1)
> * [test_events.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/dev___test_py___dev___test_events_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.dev.2)
> * [test_events.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/dev___test_py___dev___test_events_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.dev.3)
> * [test_events.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/release___test_py___release___test_events_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.release.1)
> * [test_events.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/release___test_py___release___test_events_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.release.2)
> * [test_events.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22262/testReport/junit/cluster/test_client_routes/release___test_py___release___test_events_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_events.release.3)

Those failures look like `scylla-driver` incompatibility issue i.e. the old scylla-driver was used in CI environment instead of a new one. However, I haven't found a single log that confirms it in 100%, so I added a new assert in `test_events` that explicitly checks the driver compatibility.

v6:
 - Added assert to `test_events` to check that the python driver is compatible

andrzej-jackowski-scylladb commented:
@scylladb/releng how to ensure that the newest python driver (scylla-driver 3.29.6) is used in the CI environment?
The scylladb references were already updated in https://github.com/scylladb/scylladb/pull/27359

However, I can see that the test uses an older python driver. From https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22275/testReport/junit/cluster/test_client_routes/release___test_py___release___test_events_release_1/: 
```
AssertionError: assert 'CLIENT_ROUTES_CHANGE' in frozenset({'SCHEMA_CHANGE', 'STATUS_CHANGE', 'TOPOLOGY_CHANGE'})
```

patjed41 commented on db/system_keyspace.hh:
```diff
@@ -545,6 +547,16 @@ public:
+        std::optional<int32_t> port;
+        std::optional<int32_t> tls_port;
+        std::optional<int32_t> alternator_port;
+        std::optional<int32_t> alternator_https_port;
```
I see this in the commit message:
> Ports are optional in the sense that at least one of the four must be provided.

It looks like an important invariant. I would add a comment in the struct definition.

andrzej-jackowski-scylladb replied:
added the comment

patjed41 commented on api/api-doc/client_routes.json:
```diff
@@ -0,0 +1,74 @@
+    , "/v2/client-routes":{
```
Why `/v2/`?

andrzej-jackowski-scylladb replied:
`/v2/client-routes` endpoints process arrays of objects. Handling of such structures was improved between swagger `1.2` and `2.0` versions. `get_metrics_config` and `set_metrics_config` operates on similar structures and there are also in `/v2/`.

patjed41 commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+        int retries = 10;
+        while (true) {
+            try {
+                auto batch = service::group0_batch(co_await group0_client.start_operation(ss.get_abort_source()));
```
I wonder what will happen if we try to update/delete a huge number of rows in one of those API endpoints. I see
```cpp
// If size or number of mutations is too big for raft too handle it will be
// rejected.
class group0_batch {
```
but I'm not sure what "rejected" means. Hopefully, we will get an exception (which one?) here, and this command won't be applied as a no-op unnoticed.

calebxyz commented:
> > @patjed41 we would like to push it to master as soon as possible please review it again.
> 
> My review won't be enough anyway. It's a big PR and it changes some code I don't know that well (like events sent to drivers).
> 
> We need someone from the core-frontend team to review it and ideally also @gleb-cloudius and/or @gusev-p.

@gusev-p because @gleb-cloudius doesn't work tommorow it would be great if you will review it.

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+        throw std::runtime_error("get_client_routes requires all nodes to support the CLIENT_ROUTES cluster feature");
+    }
+    co_return co_await ss.invoke_on(0, [&group0_client] (service::storage_service& ss) -> future<json::json_return_type> {
+        co_return json::json_return_type(stream_range_as_array(co_await group0_client.sys_ks().get_client_routes(), [](const db::system_keyspace::client_route & cm) {
```
The production flow is as follows:
- Cloud uses `POST` request on `/v2/client-routes` to update the client-routes information. It causes modification of `group0` table (`system.client_routes`).
- On each node, when the mutation is applied locally, the node sends an EVENT to drivers connected to this node.
- Driver obtains the new content of `system.client_routes` via CQL (i.e. `SELECT ...`).

The `GET` endpoint is currently used for testing and debugging purpose, so probably it's better that it returns the current state without waiting.

gusev-p replied:
So you have `read_barrier()` in your tests? otherwise they would be flaky.

gusev-p commented on api/api-doc/client_routes.def.json:
```diff
@@ -0,0 +1,23 @@
+    , "client_routes_entry": {
```
Something is missing in the commit description. Why the comma at the beginning? What's `client_routes.def.json`?

andrzej-jackowski-scylladb replied:
Added the explanation to the commit message.

gusev-p commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+    co_await ss.invoke_on(0, [&group0_client, &root] (service::storage_service& ss) -> future<> {
+        int retries = 10;
+        while (true) {
+            try {
```
Better to do all json parsing/validation outside of `group0_client.start_operation`, probably even on the original shard, and do the minimal conversion `const vector<client_route>& -> mutations` under the group0 lock

gusev-p commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    };
+
+                    auto mut = co_await group0_client.sys_ks().make_update_client_route_mutation(batch.write_timestamp(), route);
+                    batch.add_mutation(std::move(mut), "update_client_route");
```
This will create a weird "update_client_route; update_client_route; ..." raft command description which doesn't add much value. I doubt we even need this `group0_batch` thing.  If we extract the validation/parsing, we just need to create a `chunked_vector<canonical_mutation>` and pass it to `group0_client.prepare_command`.

andrzej-jackowski-scylladb replied:
I modified the code to remove `group0_batch` uses

gusev-p commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    }
+
+                    // Optional integer port fields; capture only those provided.
+                    std::map<sstring, std::optional<int32_t>> port_map;
```
I'm not a fan of scattering client_routes over existing "bags of everything". Can't we create a dedicated class/service for them? clients_router/manager or something. It can take an explicit constructor dependency on `raft_group0_client` in `main.cc`. We can move system_keyspace CQL-s into it to not clutter system_keyspace.hh/cc even more than they are today.

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    batch.add_mutation(std::move(mut), "update_client_route");
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
+            } catch (const ::service::group0_concurrent_modification& ex) {
```
fixed

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
+            } catch (const ::service::group0_concurrent_modification& ex) {
+                apilog.warn("Failed to set client routes due to guard conflict, retries={}", retries);
```
fixed

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+    co_return co_await ss.invoke_on(0, [&group0_client] (service::storage_service& ss) -> future<json::json_return_type> {
+        co_return json::json_return_type(stream_range_as_array(co_await group0_client.sys_ks().get_client_routes(), [](const db::system_keyspace::client_route & cm) {
+            seastar::httpd::client_routes_json::client_routes_entry obj;
+            obj.connection_id = fmt::to_string(cm.connection_id);
```
Indeed, it was a leftover from the old implementation when `connection_id` was UUID. Fixed.

andrzej-jackowski-scylladb commented on db/system_keyspace.hh:
```diff
@@ -589,6 +589,9 @@ public:
     future<mutation> make_remove_view_building_task_mutation(api::timestamp_type ts, utils::UUID id);
+    future<mutation> make_remove_client_route_mutation(api::timestamp_type ts, std::string_view connection_id, const utils::UUID& host_id);
+    future<mutation> make_update_client_route_mutation(api::timestamp_type ts, const client_route& cm);
+    future<std::vector<client_route>> get_client_routes() const;
```
fixed

andrzej-jackowski-scylladb commented on db/system_keyspace.cc:
```diff
@@ -3146,6 +3146,52 @@ future<mutation> system_keyspace::make_remove_view_building_task_mutation(api::t
+    static const sstring query = format("SELECT * from {}.{}", NAME, CLIENT_ROUTES);
+    auto rs = co_await _qp.execute_internal(query, cql3::query_processor::cache_internal::yes);
+    for (const auto& row : *rs) {
+        result.emplace_back(
```
fixed

andrzej-jackowski-scylladb commented on db/system_keyspace.cc:
```diff
@@ -110,6 +110,8 @@ namespace {
             system_keyspace::DICTS,
             system_keyspace::VIEW_BUILDING_TASKS,
+            system_keyspace::CLIENT_ROUTES,
+
```
fixed

gusev-p commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    auto mut = co_await group0_client.sys_ks().make_update_client_route_mutation(batch.write_timestamp(), route);
+                    batch.add_mutation(std::move(mut), "update_client_route");
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
```
We pass null as a value for `timeout` parameter (default constructed `std::optional<::service::raft_timeout>`), which means we would get stuck here in case of majority loss (probably will get a request timeout). Better to use `service::raft_timeout{}`.

gusev-p commented on service/raft/group0_state_machine.cc:
```diff
@@ -123,16 +123,49 @@ bool should_flush_system_topology_after_applying(const mutation& mut, const data
+    std::vector<sstring> host_ids;
+};
+
+void collect_client_routes_update(schema_ptr s, mutation& mut, storage_proxy& proxy, client_routes_update& client_routes_update) {
```
Why not return `client_routes_update` as a result of `collect_client_routes_update`? More natural than out parameter.

gusev-p commented on transport/event.hh:
```diff
@@ -79,4 +80,20 @@ public:
 
+class event::client_routes_change : public event {
+public:
+    enum class change_type { UPDATE_NODES };
```
Why declare this enum if it's not relevant to this event and is not actually used?

andrzej-jackowski-scylladb replied:
All of other existing EVENT types have their sub-types defined as a first string within an event (https://cassandra.apache.org/doc/latest/cassandra/_attachments/native_protocol_v5.html#s4.2.6). 

Currently, we have only one type of `CLIENT_ROUTES_CHANGE` event, so theoretically we could remove this field. But it can make extending the CLIENT_ROUTES_CHANGE event in the future more difficult. We have already modified this `change_type` once during the design phase (initially it had two distinct values but later we decided to merge those two types).

gleb-cloudius commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    batch.add_mutation(std::move(mut), "update_client_route");
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
+            } catch (const ::service::group0_concurrent_modification&) {
```
What is the reason to submit this with guard as opposite of using `add_entry_unguarded`? This table is never updated with RMW as far as I see. Of course it just one more prove that this should not be in group0 in the first place.

patjed41 replied:
Yes, there will be conflicts. From what I understood, the write to this table will come just after a topology operation completes, so when tablet migrations run and starve group 0 operations on followers.

However, it's supposedly fine to fail the write a few times from the cloud perspective. I don't mind using `add_entry` here because it's problematic only for the Private Link feature and has almost no impact on the rest of the system. Writes are supposed to be very rare (one in days or something) and reads don't use any barriers/guards.

Refactoring `add_entry_unguarded` would increase the scope of this already big PR with more important problems (like no driver notifications on snapshot application). If the designers of this feature are fine with conflicts that don't really affect anything else, I think it's acceptable to merge the current version of this PR.

gusev-p commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+
+                    // Optional integer port fields; capture only those provided.
+                    std::map<sstring, std::optional<int32_t>> port_map;
+                    for (const auto& name : {"port", "tls_port", "alternator_port", "alternator_https_port"}) {
```
This loop is weird, it needs several redundant allocations for the `port_map`. Better to declare a helper and call it four times.

gusev-p commented on service/raft/group0_state_machine.cc:
```diff
@@ -123,16 +123,49 @@ bool should_flush_system_topology_after_applying(const mutation& mut, const data
             auto mut = co_await to_mutation_gently(cm, s);
+
             need_system_topology_flush = need_system_topology_flush || should_flush_system_topology_after_applying(mut, proxy.data_dictionary());
+            collect_client_routes_update(s, mut, proxy, routes_update);
```
The CLIENT_ROUTES table is added to the snapshot (`storage_service::raft_pull_snapshot`)

gleb-cloudius replied:
I was asking about sending events on snapshot application, but TBH I cannot even find where it is added to the snapshot. Mind to point it out?

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+static
+future<json::json_return_type>
+rest_delete_client_routes(http_context& ctx, sharded<service::storage_service>& ss, service::raft_group0_client& group0_client, std::unique_ptr<http::request> req) {
+    if (!ss.local().get_feature_service().client_routes) {
```
Extracted the check to a helper function

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    if (!v.IsObject()) { throw bad_param_exception("Each element must be object"); }
+
+                    // Required string fields
+                    for (const auto& name : {"connection_id", "host_id", "address"}) {
```
I agree. However, the code was refactored and the `for` loop was entirely removed.

andrzej-jackowski-scylladb commented on test/cluster/test_client_routes.py:
```diff
@@ -0,0 +1,142 @@
+# This test verifies updating system to version with CLIENT_ROUTES feature in the following steps:
+# 1. Create 2 nodes with disabled CLIENT_ROUTES feature
+# 2. Verify `/v2/client-routes` rejects requests
+# 3. Enable `CLIENT_ROUTES` server after server
```
fixed

nuivall commented on main.cc:
```diff
@@ -1800,13 +1801,16 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             checkpoint(stop_signal, "initializing storage service");
             debug::the_storage_service = &ss;
+            static sharded<service::client_routes_service> client_routes;
+            client_routes.start(std::ref(stop_signal.as_sharded_abort_source()), std::ref(feature_service), std::ref(qp), std::ref(lifecycle_notifier)).get();
```
no call to client_routes.stop?

nuivall commented on service/raft/group0_state_machine.cc:
```diff
@@ -124,16 +124,47 @@ bool should_flush_system_topology_after_applying(const mutation& mut, const data
+    }
+
+    const auto pk_components = mut.decorated_key()._key.explode(*s_client_routes);
+    if (pk_components.empty()) {
```
maybe some comment what kind of case is this?

nyh replied:
> for things which we know can't happen we're typically using SCYLLA_ASSERT

This is a dangerous statement. Maybe it makes sense to use SCYLLA_ASSERT in some areas where failure is completely unrecoverable and maybe (?) it applies here, but this statement sounds very general, and I think the opposite is true in general - in general you should avoid SCYLLA_ASSERT like the plague - including for "things we know can't happen".
If you don't understand why I'm saying this, please refer to the issue I opened on this - https://github.com/scylladb/scylladb/issues/7871 - where I explained what I mean (this was already five years ago).

andrzej-jackowski-scylladb commented on service/client_routes.cc:
```diff
@@ -0,0 +1,72 @@
+future<mutation> service::client_routes_service::make_remove_client_route_mutation(api::timestamp_type ts, std::string_view connection_id, const utils::UUID& host_id) {
+    static const sstring stmt = format("DELETE FROM {}.{} WHERE connection_id = ? and host_id = ?", db::system_keyspace::NAME, db::system_keyspace::CLIENT_ROUTES);
+
+    auto muts = co_await _qp.get_mutations_internal(stmt, client_routes_query_state(), ts, {connection_id, host_id});
```
@nuivall is right, and `const utils::UUID& ` is also passed by a reference. But I think it's even better to pass `client_route_key` to make it clear rather then keeping the `string_view`.

andrzej-jackowski-scylladb commented on api/api-doc/client_routes.def.json:
```diff
@@ -0,0 +1,23 @@
+            "port": {"type": "integer"},
+            "tls_port": {"type": "integer"},
+            "alternator_port": {"type": "integer"},
+            "alternator_https_port": {"type": "integer"}
```
I think you are right. But it's property of swagger API `3.0` and we are using `2.0`. As far as I'm concerned there is no similar mechanism in `v2` api.

andrzej-jackowski-scylladb commented on service/raft/group0_state_machine.cc:
```diff
@@ -124,16 +124,47 @@ bool should_flush_system_topology_after_applying(const mutation& mut, const data
+
+future<> write_mutations_to_database(storage_service& storage_service, storage_proxy& proxy, gms::inet_address from, utils::chunked_vector<canonical_mutation> cms) {
     utils::chunked_vector<frozen_mutation_and_schema> mutations;
+    std::vector<client_routes_service::client_route_key> routes_update;
```
Yes, especially if you see the implementation of sending events on snapshot transfers it's much easier like that.

andrzej-jackowski-scylladb commented on api/storage_service.cc:
```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                    continue;
+                }
+                throw;
+            }
```
Cloud scripts (the API user) will also implement retries on their side, so adding backoff in core seems unnecessary.

andrzej-jackowski-scylladb commented on test/cluster/test_client_routes.py:
```diff
@@ -0,0 +1,142 @@
+
+    wait_for(client_routes_ready, time.time() + 10)
+
+
```
I agree it's a basic property of group0. However, I think it's better to test the scenario mentioned by Petr because in almost all tests `client_routes` are empty. I added `test_client_routes_node_restart` to increase the coverage and avoid regression in the future.

patjed41 commented on service/client_routes.hh:
```diff
@@ -0,0 +1,65 @@
+
+    struct client_route_key {
+        // Avoid sstring because client_route_key is passed to `invoke_all`
+        std::string connection_id;
```
What's the problem with passing `sstring` to `invoke_on_all`?

patjed41 commented on service/client_routes.cc:
```diff
@@ -72,3 +72,57 @@ future<std::vector<service::client_routes_service::client_route_entry>> service:
+    utils::chunked_vector<canonical_mutation> cmuts;
+
+    for (auto& entry : route_entries) {
+        auto mut = co_await make_update_client_route_mutation(guard.write_timestamp(), entry);
```
Not related to the current version of this PR, but I wonder what timestamp would we pass here if we used `add_entry_unguarded`.

patjed41 commented on test/cluster/test_client_routes.py:
```diff
@@ -0,0 +1,161 @@
+        if len(res) == expected_size:
+            return res
+        return None
+    await wait_for(lambda: expected_client_routes_size(cql, expected_routes_size), time.time() + 10)
```
You can harden these tests a bit by waiting until all live nodes have the expected `client_routes` size. You can pass a list of hosts to `wait_for_expected_client_routes_size` and execute the statement for each host.

patjed41 commented on test/cluster/test_client_routes.py:
```diff
@@ -0,0 +1,161 @@
+
+async def wait_for_expected_client_routes_size(cql, expected_routes_size):
+    async def expected_client_routes_size(cql, expected_size):
+        res = list(cql.execute("SELECT * FROM system.client_routes"))
```
`cql.run_async` should a better choice in a async function.

patjed41 commented on api/client_routes.cc:
```diff
@@ -127,7 +127,7 @@ static std::vector<service::client_routes_service::client_route_key> parse_delet
 static
 future<json::json_return_type>
-rest_delete_client_routes(http_context& ctx, sharded<service::client_routes_service>& cr,  std::unique_ptr<http::request> req) {
+rest_delete_client_routes(http_context& ctx, sharded<service::client_routes_service>& cr, std::unique_ptr<http::request> req) {
```
nit: This change shouldn't be in the "service: transport: add CLIENT_ROUTES_CHANGE event" commit.

patjed41 commented on test/cluster/test_client_routes.py:
```diff
@@ -159,3 +160,115 @@ async def fail_req(f):
+
+    await manager.server_start(server_to_restart.server_id)
+    cql = await manager.get_cql_exclusive(server_to_restart)
+    await wait_for_expected_client_routes_size(cql, 0)
```
This check looks unreliable.  It could happen that we read from `client_routes` too quickly, before a snapshot/command is (unexpectedly) applied. I would wait for a log printed after hitting wait_transfer_snapshot.

We should also verify that this test does what we expect. The test could be broken (by e.g. misusing trigger_snapshot) and `client_routes` could be unexpectedly updated in a group 0 command, not snapshot. We should check that this update really happens in transfer_snapshot. An easy way to do this is by adding a trace log there for non-empty client_routes updates and looking for it in the test.

patjed41 reviewed (commented):
Some of my comments are cosmetic and I won't block this PR because of them. Please first respond to:
- https://github.com/scylladb/scylladb/pull/27323#discussion_r2614615138
- https://github.com/scylladb/scylladb/pull/27323#discussion_r2615042405

Even if we merge this PR as it is, I think we should still refactor `add_entry_unguarded` and use it in `client_routes_service`. Let's open a follow-up issue once this PR is merged.

There is a conflict to resolve.

andrzej-jackowski-scylladb commented on api/client_routes.cc:
```diff
@@ -0,0 +1,182 @@
+        throw bad_param_exception(fmt::format("'{}' must be an integer", name));
+    }
+    auto port = it->value.GetInt();
+    if (port < 0 || port > 65535) {
```
I checked and public clouds typically doesn't allow 0 a port on the load balancer. Therefore, I agree it's better to not allow 0. Changed the `if` and related tests.

andrzej-jackowski-scylladb commented:
v13:
 - Rebase to resolve conflicts

v14:
 - Changed port verification to disallow port = 0
 - Restored the monkeypatching of python-driver as https://github.com/scylladb/scylladb/pull/27567 hasn't been merged yet. 
 - Renamed injected error, improved tests
 - Changed std::string to sstring
 - Moved `invoke_on(0, ...` to `client_route_service`

patjed41 commented on service/raft/group0_state_machine.cc:
```diff
@@ -451,11 +490,22 @@ future<> group0_state_machine::transfer_snapshot(raft::server_id from_id, raft::
+                    collect_client_routes_update(mut, client_routes_update);
+                }
+            }
+        }
```
We always send the event when `client_routes` is non-empty, so if the node that is applying the snapshot has the latest state of `client_routes`, the event will be send unnecessarily. However, snapshot transfer is rare, so this shouldn't be an issue. Also, I have no idea how to fix this. It looks like snapshot transfer isn't made to react on new changes.
